hal/lib/athena/Relay.cpp

```diff
@@ -8,111 +8,125 @@
+
+  int16_t pin = getPortHandlePin(port_handle);
+  if (pin == InvalidHandleIndex) {
     *status = PARAMETER_OUT_OF_RANGE;
-    return false;
-  } else {
-    return true;
+    return HAL_INVALID_HANDLE;
   }
-}
 
-/**
- * Set the state of a relay.
- * Set the state of a relay output to be forward. Relays have two outputs and
- * each is
- * independently set to 0v or 12v.
- */
-void setRelayForward(void* digital_port_pointer, bool on, int32_t* status) {
-  DigitalPort* port = (DigitalPort*)digital_port_pointer;
-  if (!verifyRelayChannel(port, status)) {
-    return;
+  if (!fwd) pin += kRelayHeaders;  // add 4 to reverse pins
+
+  auto handle = relayHandles.Allocate(pin, status);
+
+  if (*status != 0)
+    return HAL_INVALID_HANDLE;  // failed to allocate. Pass error back.
+
+  auto port = relayHandles.Get(handle);
+  if (port == nullptr) {  // would only occur on thread issue.
+    *status = PARAMETER_OUT_OF_RANGE;
+    return HAL_INVALID_HANDLE;
   }
 
-  {
-    std::lock_guard<priority_recursive_mutex> sync(digitalRelayMutex);
-    uint8_t forwardRelays = relaySystem->readValue_Forward(status);
-    if (on)
-      forwardRelays |= 1 << port->pin;
-    else
-      forwardRelays &= ~(1 << port->pin);
-    relaySystem->writeValue_Forward(forwardRelays, status);
+  if (!fwd) {
+    pin -= kRelayHeaders;  // subtract number of headers to put pin in range.
+    port->fwd = false;     // set to reverse
+  } else {
+    port->fwd = true;  // set to forward
   }
+
+  port->pin = static_cast<uint8_t>(pin);
+  return handle;
+}
+
+void freeRelayPort(HalRelayHandle relay_port_handle) {
+  // no status, so no need to check for a proper free.
+  relayHandles.Free(relay_port_handle);
+}
+
+bool checkRelayChannel(uint8_t pin) {
+  return pin < kRelayHeaders;  // roboRIO only has 4 headers, and the FPGA has
+                               // seperate functions for forward and reverse,
+                               // instead of seperate pin IDs
 }
 
 /**
  * Set the state of a relay.
- * Set the state of a relay output to be reverse. Relays have two outputs and
- * each is
- * independently set to 0v or 12v.
+ * Set the state of a relay output.
  */
-void setRelayReverse(void* digital_port_pointer, bool on, int32_t* status) {
-  DigitalPort* port = (DigitalPort*)digital_port_pointer;
-  if (!verifyRelayChannel(port, status)) {
+void setRelay(HalRelayHandle relay_port_handle, bool on, int32_t* status) {
+  auto port = relayHandles.Get(relay_port_handle);
+  if (port == nullptr) {
+    *status = PARAMETER_OUT_OF_RANGE;
     return;
   }
-
-  {
-    std::lock_guard<priority_recursive_mutex> sync(digitalRelayMutex);
-    uint8_t reverseRelays = relaySystem->readValue_Reverse(status);
-    if (on)
-      reverseRelays |= 1 << port->pin;
-    else
-      reverseRelays &= ~(1 << port->pin);
-    relaySystem->writeValue_Reverse(reverseRelays, status);
+  std::lock_guard<priority_recursive_mutex> sync(digitalRelayMutex);
+  uint8_t relays = 0;
```

Comment:
I'm assuming so. The reason for the lock is so it can read, update the byte mask, and then rewrite. And there's no lock in the get function. When I was talking to Joe at worlds he said all of that should be thread safe.